### PR TITLE
Update incorrect caps instead recreating whole client key, (bsc#961108)

### PR DESCRIPTION
**Pull request by mjura**
cherry-picked from commit
https://github.com/crowbar/crowbar-ceph/commit/64a4c30b2d33533a3a360184dcff743c5a60a446

### CODE REVIEW

**Comment by dirkmueller** [comment]
@mjura can you add the link to the master id ? this cherry-pick id doesn't make sense.

**Comment by mjura** [comment]
@dirkmueller : Sure, I've changed commit message

File: chef/cookbooks/ceph/providers/client.rb [hunk 1]
```diff
@@ -61,23 +61,23 @@ def get_caps(ceph_conf, admin_keyring, keyname)
...
-    get_or_create = Mixlib::ShellOut.new(cmd)
-    get_or_create.run_command
-  end
-  get_or_create.error!
+  return get_or_create.error! unless get_or_create.stderr.scan(/EINVAL.*but cap.*does not match/)
+  Chef::Log.info("Updating incorrect caps for #{keyname}")
+  # don't remove custom existing caps for osd pools
+  cur_caps = get_caps(ceph_conf, admin_keyring, keyname)
+  caps_osd = caps["osd"]
+  caps_osd.insert(0, cur_caps["osd"] + ",") unless cur_caps.empty?
```
Comment (vuntz) [inline chef/cookbooks/ceph/providers/client.rb]: In the PR for crowbar-ceph, you use `if cur_caps["osd"] && !cur_caps["osd"].empty?`

File: chef/cookbooks/ceph/providers/client.rb [hunk 2]
```diff
@@ -61,23 +61,23 @@ def get_caps(ceph_conf, admin_keyring, keyname)
...
+  caps["osd"] = caps_osd.split(",").collect(&:strip).uniq.join(", ")
```
Comment (vuntz) [inline chef/cookbooks/ceph/providers/client.rb]: In the PR for crowbar-ceph, you have a `sort` before `uniq`.

**Comment by vuntz** [comment]
This needs to be updated to latest code from https://github.com/crowbar/crowbar-ceph/pull/32

**Comment by vuntz** [comment]
+1

File: chef/cookbooks/ceph/providers/client.rb (hunk 2 above)
Comment (nicolasbock) [inline chef/cookbooks/ceph/providers/client.rb]: As far as I understand it, the uniq method uses the hashes of the list object to uniquify the list, and sorting it beforehand is not necessary.

File: chef/cookbooks/ceph/providers/client.rb [hunk 3]
```diff
@@ -61,23 +61,30 @@ def get_caps(ceph_conf, admin_keyring, keyname)
...
-    get_or_create = Mixlib::ShellOut.new(cmd)
-    get_or_create.run_command
+  return get_or_create.error! unless get_or_create.stderr.scan(/EINVAL.*but cap.*does not match/)
+  Chef::Log.info("Updating incorrect caps for #{keyname}")
+  # create list of uniq pools
+  pools = caps["osd"].split(",").collect(&:strip).map { |x| x.gsub(/allow [rwx]{1,3} /, "") }.uniq
+  # get current existing capabilities
+  cur_caps = get_caps(ceph_conf, admin_keyring, keyname)
+  if cur_caps["osd"] && !cur_caps["osd"].empty?
+    cur_caps["osd"].split(",").collect(&:strip).sort.uniq.each do |cap|
```
Comment (nicolasbock) [inline chef/cookbooks/ceph/providers/client.rb:77]: From my reading of ruby Array methods, you don't have to `sort` the Array before `uniq`. The result should be the same though, with or without `sort`.

**Comment by nicolasbock** [comment]
:+1:

File: chef/cookbooks/ceph/providers/client.rb (hunk 3 above)
Comment (dirkmueller) [reply chef/cookbooks/ceph/providers/client.rb:77]: Right, but the reason sort is there becasue the result needs to be "stable" independent of arbitrary dict reorderings between different runs. theoretically .uniq.sort would be slightly more efficient but I guess it does not matter whatsoever.

**Review by dirkmueller — APPROVED** [review]

File: chef/cookbooks/ceph/providers/client.rb (hunk 3 above)
Comment (nicolasbock) [reply chef/cookbooks/ceph/providers/client.rb:77]: ah, thanks, that makes sense.

**Comment by nicolasbock** [comment]
+1

**Review by nicolasbock — APPROVED** [review]
Giving it another +1

**Comment by mjura** [comment]
Can be closed

**Pull request closed without merging**